### ISSUE TITLE
is `yolo detect val` running `test` or `valid` from data.yml?

### ISSUE THREAD

**Issue by FrancescoSaverioZuppichini** [ultralytics/ultralytics#523]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Like title

Thanks :)

Fra

**Comment by AyushExel** [comment]
It looks for `val` first and then for `test` if val is not found.

**Comment by FrancescoSaverioZuppichini** [comment]
Thanks @AyushExel , that's very confusing. What is the correct way to run the model on the `test`?

**Comment by AyushExel** [comment]
`model.val(data="data.yaml"`
The dataset should have the same structure as a yolov5 dataset. it looks for `val` split. If that is not found, it looks for `test` as a fallback.
Is it clear now?

**Comment by FrancescoSaverioZuppichini** [comment]
No, it is not, let me try to rephase it and apologies if I was not clear before.

How do I evaluate the `test` split from `data.yml` using the `cli`

**Comment by AyushExel** [comment]
@FrancescoSaverioZuppichini oh i see. No we don't support passing specific `splits` to validate. But it does seem like a good idea.
Most of what is supported now should be pretty much the same as yolov5.

**Comment by FrancescoSaverioZuppichini** [comment]
so in `yolov5` when I run the `python val.py --data $dataset/data.yaml` - that was running on the `val` split not on `test`?

**Comment by AyushExel** [comment]
@FrancescoSaverioZuppichini yes it should by default. these 2 lines might make things clear. 
https://github.com/ultralytics/yolov5/blob/master/val.py#L106
https://github.com/ultralytics/yolov5/blob/master/val.py#L173

**Comment by FrancescoSaverioZuppichini** [comment]
@AyushExel that's very bad news for me since all RF100 benchmarks now are invalid - they run on the validation set! :cry: 

Thanks a lot!

**Comment by AyushExel** [comment]
@FrancescoSaverioZuppichini 😭 
If it helps, I think all yolov5 derivative works should be doing the same as they use the same yaml structure. So at least your study should still be comparable and not completely invalid. But I'm not sure about this.

**Comment by FrancescoSaverioZuppichini** [comment]
My fix is to replace the `val` entry in the `yaml` file with the path to the `test` set

**Comment by AyushExel** [comment]
@glenn-jocher what do you think about supporting custom splits for validation ?
``` 
yolo detect val data=data.yaml split=test_custom
```
I think this could be a useful addition. We can push it in next update PR I'm working on here https://github.com/ultralytics/ultralytics/pull/580

**Comment by geezacoleman** [comment]
In v5 it's possible to explicitly use the test split by setting `task = 'test'`, is that correct? In v5 when I run `val.py --task 'test'` it draws from `../labels/test` with a different number of instances to the `val` set. Is this possible in v8? 

It's fairly useful to have this capability for research. It sounds like the workaround for this is by renaming test to val in all the `yaml` files, but with 5 fold CV and multiple comparisons it can be quite laboursome. Thanks for working on this.

**Comment by heitorrapela** [comment]
> My fix is to replace the `val` entry in the `yaml` file with the path to the `test` set

This is not a good idea, as your are going to use this metric to select the hyperparameters and early stopping.

**Comment by glenn-jocher** [comment]
@heitorrapela That's one way to do it. Just remember if you do decide to use the `test` set in place of the `val` set for any reason, you're essentially using your test data for validation. This could lead to overfitting on the test set and your results may not generalize well to new data. 

Furthermore, this 'fix' might affect your model's generalization as we use validation metrics for hyperparameter tuning and early stopping. Therefore, it's crucial that validation and test sets remain separate to ensure unbiased evaluation of the model.

For research or testing purposes, you can create a separate run where you replace the `val` entry with your `test` set path. However, make sure to revert back to the original setting for actual model training to maintain the integrity of your validation and test sets.

Let me know if you have additional questions or if there's anything else you'd like to discuss!
